### PR TITLE
[Fix] No training log when the num of iterations is smaller than the default interval

### PR DESCRIPTION
**Pull request by shufanwu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

No training log when the num of iterations is smaller than the default interval. [https://github.com/open-mmlab/mmengine/issues/935](Issue)

## Modification

Add judgement in  function ```after_train_iter``` of class ```LoggerHook```.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by HAOCHENYE — COMMENTED** [review]
Hi, thanks for your contribution. The modification makes sense to me, but should we also update the `after_val_iter` and `after_test_iter`?

**Comment by shufanwu** [comment]
> Hi, thanks for your contribution. The modification makes sense to me, but should we also update the `after_val_iter` and `after_test_iter`?
I think we don't need to  update these because in val or test task, the loggerhook `after_val_epoch` and  `after_test_epoch` will be called after each epoch.

**Comment by HAOCHENYE** [comment]
> > Hi, thanks for your contribution. The modification makes sense to me, but should we also update the `after_val_iter` and `after_test_iter`?
> 
> I think we don't need to update these because in val or test task, the loggerhook `after_val_epoch` and `after_test_epoch` will be called after each epoch.

It makes sense to me :+1:

File: mmengine/hooks/logger_hook.py [hunk 1]
```diff
@@ -167,7 +167,8 @@ def after_train_iter(self,
             tag, log_str = runner.log_processor.get_log_after_iter(
                 runner, batch_idx, 'train')
         elif (self.end_of_epoch(runner.train_dataloader, batch_idx)
-              and not self.ignore_last):
+              and (not self.ignore_last
+                   or len(runner.train_dataloader) < self.interval)):
```
Comment (HAOCHENYE) [inline mmengine/hooks/logger_hook.py:171]:
```suggestion
                   or len(runner.train_dataloader) <= self.interval)):
```

**Comment by HAOCHENYE** [comment]
Hi! Could you add a unit test for `LoggerHook` to validate that the logs will be output when the length of the train dataloader is smaller than the log interval?

**Comment by HAOCHENYE** [comment]
Hi, it could be more proper to add unit test [here](https://github.com/open-mmlab/mmengine/blob/6b366f236c93d2735b336ee41dc521db1dd370e0/tests/test_hooks/test_logger_hook.py#L64), and we should test runner.logger.info will be called when the length of dataloader is smaller than the log interval.

File: tests/test_runner/test_runner.py [hunk 2]
```diff
@@ -1727,6 +1727,13 @@ def train_step(self, *args, **kwargs):
         with self.assertRaisesRegex(AssertionError, 'If you want to validate'):
             runner.train()
 
+        # 13 Test the logs will be output when the length of
+        # the train dataloader is smaller than the log interval
+        cfg = copy.deepcopy(self.epoch_based_cfg)
+        cfg.experiment_name = 'test_train13'
+        runner = Runner.from_cfg(cfg)
+        runner.train()
```
Comment (zhouzaida) [inline tests/test_runner/test_runner.py:1735]:
```suggestion
        # 13 Test the logs will be printed when the length of
        # train_dataloader is smaller than the interval set in LoggerHook
        cfg = copy.deepcopy(self.epoch_based_cfg)
        cfg.experiment_name = 'test_train13'
        cfg.default_hooks = dict(logger=dict(type='LoggerHook', interval=5))
        runner = Runner.from_cfg(cfg)
        runner.train()
        with open(runner.logger._log_file, 'r') as f:
            log = f.read()
        self.assertIn('Epoch(train) [1][4/4]', log)
```

**Review by zhouzaida — APPROVED** [review]

File: tests/test_runner/test_runner.py [hunk 3]
```diff
@@ -1727,6 +1727,17 @@ def train_step(self, *args, **kwargs):
...
             runner.train()
 
+        # 13 Test the logs will be printed when the length of
+        # train_dataloader is smaller than the interval set in LoggerHook
+        cfg = copy.deepcopy(self.epoch_based_cfg)
+        cfg.experiment_name = 'test_train13'
+        cfg.default_hooks = dict(logger=dict(type='LoggerHook', interval=5))
+        runner = Runner.from_cfg(cfg)
+        runner.train()
+        with open(runner.logger._log_file, 'r') as f:
```
Comment (zhouzaida) [inline tests/test_runner/test_runner.py:1737]:
```suggestion
        with open(runner.logger._log_file) as f:
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**